testlapack/matgen.go

```diff
@@ -183,6 +183,92 @@ func Dlagsy(n int, d []float64, a []float64, lda int, rnd *rand.Rand, work []flo
 	}
 }
 
+// Dlagge generates a real general m×n matrix A, by pre- and post-multiplying
+// a real diagonal matrix D with random orthogonal matrices:
+//  A = U*D*V.
+//
+// d must have length min(m,n), and work must have length m+n, otherwise Dlagge
+// will panic.
+//
+// The parameters ku and kl are unused.
+func Dlagge(m, n, kl, ku int, d []float64, a []float64, lda int, rnd *rand.Rand, work []float64) {
+	checkMatrix(m, n, a, lda)
+	if len(d) != min(m, n) {
+		panic("testlapack: bad length of d")
+	}
+	if len(work) < m+n {
+		panic("testlapack: insufficient work length")
+	}
+
+	// Initialize A to diagonal matrix.
+	for i := 0; i < m; i++ {
+		for j := 0; j < n; j++ {
+			a[i*lda+j] = 0
+		}
+	}
+	for i := 0; i < min(m, n); i++ {
+		a[i*lda+i] = d[i]
+	}
+
+	// Quick exit if the user wants a diagonal matrix.
+	// if kl == 0 && ku == 0 {
+	// 	return
+	// }
+
+	bi := blas64.Implementation()
+
+	// Pre- and post-multiply A by random orthogonal matrices.
+	for i := min(m, n) - 1; i >= 0; i-- {
+		if i < m-1 {
+			for j := 0; j < m-i; j++ {
+				work[j] = rnd.NormFloat64()
+			}
+			wn := bi.Dnrm2(m-i, work[:m-i], 1)
+			wa := math.Copysign(wn, work[0])
+			var tau float64
+			if wn != 0 {
+				wb := work[0] + wa
+				bi.Dscal(m-i-1, 1/wb, work[1:m-i], 1)
+				work[0] = 1
+				tau = wb / wa
+			}
+
+			// Multiply A[i:m,i:n] by random reflection from the left.
+			bi.Dgemv(blas.Trans, m-i, n-i,
+				1, a[i*lda+i:], lda, work[:m-i], 1,
+				0, work[m:m+n-i], 1)
+			bi.Dger(m-i, n-i,
```

Comment:
Do you have any suggestion? I'm trying to be consistent here: 1st line - matrix size, 2nd line - operation (e.g., scalar * mat * vec or scalar * vec * vec^T), 3rd line - destination.